module/VuFind/src/VuFind/Config/Feature/ConfigSettingPropertiesTrait.php

```diff
@@ -0,0 +1,252 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see
+ * <https://www.gnu.org/licenses/>.
+ *
+ * @category VuFind
+ * @package  Config
+ * @author   Aleksi Peebles <aleksi.peebles@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+
+namespace VuFind\Config\Feature;
+
+use VuFind\Exception\BadConfig;
+
+use function array_key_exists;
+use function in_array;
+use function is_array;
+
+/**
+ * Trait implementing SettingPropertiesInterface
+ *
+ * @category VuFind
+ * @package  Config
+ * @author   Aleksi Peebles <aleksi.peebles@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+trait ConfigSettingPropertiesTrait
+{
+    /**
+     * Required and conditionally required settings.
+     *
+     * @var array
+     */
+    protected array $requiredSettings = [self::DEFAULT_CONTEXT => []];
+
+    /**
+     * Localizable settings.
+     *
+     * @var array
+     */
+    protected array $localizableSettings = [self::DEFAULT_CONTEXT => []];
+
+    /**
+     * Return required and conditionally required settings.
+     *
+     * @param string $contextKey Key identifying the context (optional)
+     *
+     * @return array
+     */
+    public function getRequiredSettings(
+        string $contextKey = self::DEFAULT_CONTEXT
+    ): array {
+        return $this->requiredSettings[$contextKey];
+    }
+
+    /**
+     * Is the setting required?
+     *
+     * The optional context and context key parameters are used to evaluate if a
+     * conditionally required setting is required. If context is omitted returns
+     * true for both required and conditionally required settings.
+     *
+     * @param string $setting    Setting
+     * @param array  $context    Settings to be used in evaluation (optional)
+     * @param string $contextKey Key identifying the context (optional)
+     *
+     * @return bool
+     */
+    public function isRequiredSetting(
+        string $setting,
+        array $context = [],
+        string $contextKey = self::DEFAULT_CONTEXT
+    ): bool {
+        // Default implementation does not handle conditional requirements.
+        return in_array($setting, $this->getRequiredSettings($contextKey));
+    }
+
+    /**
+     * Return settings that may be localized.
+     *
+     * @param string $contextKey Key identifying the context (optional)
+     *
+     * @return array
+     */
+    public function getLocalizableSettings(
+        string $contextKey = self::DEFAULT_CONTEXT
+    ): array {
+        return $this->localizableSettings[$contextKey];
+    }
+
+    /**
+     * Is the setting localizable?
```

Comment:
Maybe we should define what this means in the comments. At first glance, I assumed that this meant it would be run through the translator, but I see now that it actually means that the setting can be represented as an array keyed by locale code.